### PR TITLE
Review: Swarm integration

### PR DESCRIPTION
**Pull request by wallnerryan**
Looking to review this, few things
- This integrates installing Swarm with the `--swarm` flag to `uft-flocker-config` script
- The integration only works with swarm and ubuntu
- The integration used default ports `2375` for swarm agents and `2357` for swarm master. Should this be configurable? What about TLS?
- This restarts the docker daemon within a UFT cluster to listen on unix socket and external port, should we edit the `docker.service` file (managed by systemd) instead?
- needs to be tested with one test.

### CODE REVIEW

File: unofficial_flocker_tools/config.py [hunk 1]
```diff
@@ -143,6 +144,36 @@ def main(reactor, args):
...
+    if c.config["os"] == "ubuntu":
+        if len(sys.argv) > 2:
+            if sys.argv[2] == "--swarm":
+                clusterid = c.runSSH(c.config["control_node"], """
+docker run swarm create""").strip()
+                log("Created Swarm ID")
+                for node in c.config["agent_nodes"]:
+                    d = c.runSSHAsync(node['public'], """
+service docker stop
+docker daemon -H unix:///var/run/docker.sock -H tcp://0.0.0.0:2375 >> /tmp/dockerlogs 2>&1 &
```
Comment (lukemarsden) [inline unofficial_flocker_tools/config.py:156]: Let's regex `/etc/default/docker` with `sed` instead. This is an ephemeral, unreliable way of starting docker with some args.

File: unofficial_flocker_tools/config.py [hunk 2]
```diff
@@ -143,6 +144,36 @@ def main(reactor, args):
 
     yield gatherResults(deferreds)
 
+    if c.config["os"] == "ubuntu":
+        if len(sys.argv) > 2:
+            if sys.argv[2] == "--swarm":
```
Comment (lukemarsden) [inline unofficial_flocker_tools/config.py:149]: Probably worth introducing `twisted.usage.Options` for proper args parsing pretty soon, if not immediately. See https://github.com/ClusterHQ/unofficial-flocker-tools/blob/master/unofficial_flocker_tools/flocker_volumes.py#L83 and friends for sample `optArgs` usage.

File: unofficial_flocker_tools/config.py [hunk 3]
```diff
@@ -143,6 +144,36 @@ def main(reactor, args):
...
+                clusterid = c.runSSH(c.config["control_node"], """
+docker run swarm create""").strip()
+                log("Created Swarm ID")
+                for node in c.config["agent_nodes"]:
+                    d = c.runSSHAsync(node['public'], """
+service docker stop
+docker daemon -H unix:///var/run/docker.sock -H tcp://0.0.0.0:2375 >> /tmp/dockerlogs 2>&1 &
+""")
+                    # Let daemon come up
+                    time.sleep(3)
```
Comment (lukemarsden) [inline unofficial_flocker_tools/config.py:159]: Would be nice if we could `loop_until` the daemon is reachable, that would be more resilient.

**Comment by lukemarsden** [comment]
Generally this looks great, thanks for working on this! I've left a few review comments which should be addressed before we merge this. I'd be glad to pair fixup-and-merge this changeset with you.

File: unofficial_flocker_tools/config.py (hunk 1 above)
Comment (wallnerryan) [inline unofficial_flocker_tools/config.py:156]: agreed, that is the way I ultimately wanted to do it. For sake of time I did it thise way but agree setting the options in default/docker its better. thanks will work on this.

File: unofficial_flocker_tools/config.py (hunk 3 above)
Comment (wallnerryan) [inline unofficial_flocker_tools/config.py:159]: okay, makes sense

File: unofficial_flocker_tools/config.py (hunk 2 above)
Comment (wallnerryan) [inline unofficial_flocker_tools/config.py:149]: will look into this

**Comment by lukemarsden** [comment]
I'm going to merge this now because having _some_ (even undocumented) support for swarm in UFT is better than nothing. But I'd still really like the review comments to be addressed in a follow-up PR. @wallnerryan can you make sure that's on your (not insignificant, I'm sure) todo list please? :grinning:

**Comment by wallnerryan** [comment]
will take a look at addressing these changes soon. Will also update https://github.com/ClusterHQ/unofficial-flocker-tools/issues/48 when it gets completed.

**Pull request merged**